---
title: "Generic microbiome analysis"
subtitle: "03 Taxonomical analysis"
description: "V1.2"   
author: "Roey Angel"
date: "2021-10-14"
bibliography: references.bib
link-citations: yes
csl: fems-microbiology-ecology.csl
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    keep_md: true
    number_sections: false
    highlight: "pygments"
    theme: "flatly"
    dev: "png"
    df_print: "kable"
    fig_caption: true
    code_folding: "show"
---






[roey.angel@bc.cas.cz](mailto: roey.angel@bc.cas.cz)  

## Taxonomical analysis
This analysis explores the taxonomical distribution patters in the different samples, based on the DADA2-produced sequences. Large parts of this script are based on [this protocol](https://f1000research.com/articles/5-1492/v2) and the accompanying publication by Callahan and colleagues [-@callahan_bioconductor_2016].

### Setting general parameters:

```r
set.seed(1000)
min_lib_size <- 2500
data_path <- "./DADA2_FALSE/"
Metadata_table <- "./AMetadata_decontam.csv"
Seq_table <- "DADA2.seqtab_nochim_decontam.tsv"
Tax_table <- "DADA2.taxa_silva.tsv"
Proj_name <- "Generic"
Ps_file <- paste0(Proj_name, "_decontam.Rds")
Seq_file <- "DADA2.reps_decontam.fa"
Var1 = "Density (g ml-1)" # e.g sampling point / replicate
Var2 = "Treatment" # e.g. a treatment or a manipulation
Var3 <- "Label..18O." # e.g. a treatment/manipulation or an important covariant
Var4 = "" # e.g. an important covariant
```

### Load phyloseq object
Read a decontaminated phyloseq object generated by [02_Decontamination.html](02_Decontamination.html). 
Alternatively read raw data and generate a phyloseq object if 02_Decontamination.html was not run.


```r
# # Uncomment if Decontamination wasn't run
# 
# # read OTU mat from data file
# read_tsv(paste0(data_path, Seq_table),
#          trim_ws = TRUE) %>%
#   rename_with(., ~ str_remove(.x, "_L001")) %>%
#   identity() ->
#   abundance_mat # in tibble format
# 
# # get short names of samples
# # abundance_mat %>%
# #   rownames() %>%
# #   str_remove("^Roey[0-9]{3,}-?") %>%
# #   str_split("_", simplify = T) %>%
# #   .[, 1] ->
# #   short_names
# 
# # Read metadata file
# read_csv(Metadata_table,
#                         trim_ws = TRUE) %>%
#   mutate(`ng DNA g-1` = replace(`ng DNA g-1`, which(`ng DNA g-1` == 0 | is.na(`ng DNA g-1`)), 1)) %>%  # add pseudo count
#   filter(str_remove(Read1_file, "_R1_001_noPrimers.fastq.gz") %in% colnames(abundance_mat)) %>% # remove metadata rows if the samples did not go through qual processing
#   mutate(`Library size` = colSums(select(abundance_mat, -ASV))) %>% # Add lib size
#   mutate(to_names = str_remove(Read1_file, "_R1_001_noPrimers.fastq.gz"), .before = 1) %>%
#   mutate(across(c(!!sym(Var1),
#                   !!sym(Var2),
#                   !!sym(Var3),
#                   !!sym(Var4)), ~factor(.))) %>%
#   # mutate(Identifier = paste(Site,
#   #                           `Plant species`,
#   #                           `Preservation method`,
#                             # Replicate,
#   #                           sep = "_"))  %>%   # optional. For merging samples
#   identity() ->
#   Metadata
# 
# # Order abundance_mat samples according to the metadata
# sample_order <- match(Metadata$to_names, str_remove(colnames(select(abundance_mat, -ASV)), "_L001"))
# abundance_mat %<>% select(c(1, sample_order + 1))
# 
# # read taxonomy from data file
# Raw_tax_data <- read_tsv(paste0(data_path, Tax_table),
#                         trim_ws = TRUE, col_names = TRUE)
# Raw_tax_data %<>%
#   replace(., is.na(.), "Unclassified")
# 
# # Potentially store tax classification BS values
# # Raw_tax_data %>%
# #   dplyr::select(.,
# #          `Kingdom (BS)`,
# #          `Phylum (BS)`,
# #          `Class (BS)`,
# #          `Order (BS)`,
# #          `Family (BS)`,
# #          `Genus (BS)`) %>%
# #   cbind(Name = colnames(abundance_mat),. ) ->
# #   Taxonomy.bs
# 
# # merge it downstream with the PS object
# # taxTraits <- tax_table(cbind(tax_table(Ps_obj), taxTraits))
# # Ps_obj <- merge_phyloseq(Ps_obj, taxTraits)
# # Taxonomy.bs %<>%
# #   filter(Taxonomy.bs$Name %in% row.names(Ps_obj_filt@tax_table))
# 
# Raw_tax_data %>%
#   dplyr::select(.,
#          Kingdom,
#          Phylum,
#          Class,
#          Order,
#          Family,
#          Genus) %>%
#   # map_dfr(., as_factor) %>%
#   # map_dfr(fct_expand, "Rare")  %>%
#   mutate(to_names = abundance_mat$ASV, .before = 1)-> # must be a matrix or phyloseq drops row names and gives and error
#   Taxonomy
# # row.names(Taxonomy) <- colnames(abundance_mat)
# # colnames(Taxonomy) <-
# #   c("Domain", "Phylum", "Class", "Order", "Family", "Genus")
# 
# # read sequence data
# ASV_seqs <- readDNAStringSet(
#   file = paste0(data_path, Seq_file),
#   format = "fasta",
#   nrec = -1L,
#   skip = 0L,
#   seek.first.rec = FALSE,
#   use.names = TRUE)
# 
# # generate phyloseq object. **Note: only speedyseq allows constructing phyloseq from tibble!**
# Ps_obj <- phyloseq(otu_table(abundance_mat, taxa_are_rows = TRUE),
#                    sample_data(Metadata),
#                    tax_table(Taxonomy),
#                    refseq(ASV_seqs))
# # rename_with_sample_data(Ps_obj, colnames(select(Metadata, -to_names)))
# 
# # merge samples in case the company re-sequenced some samples. **Note! Don't use it now to merge real DNA samples. Instead, do it downstream in one of the next scripts.**
# Ps_obj %>%
#   phyloseq_merge_samples("Read1_file") %>%
#   filter_taxa(., function(x) sum(x) > 0, TRUE) ->
#   Ps_obj_merged

# Comment if there's no RDS file
readRDS(paste0(data_path, Ps_file)) ->
  Ps_obj
Ps_obj %<>% 
  subset_samples(., !grepl(paste(c("CTRL", "NTC", "mock", "blank"), collapse = "|"), sample_names(Ps_obj))) 

# Reorder factors for plotting
#sample_data(Ps_obj)$Spill %<>% fct_relevel("Old", after = Inf)
```

### Exploring dataset features

#### Taxa-based filtering 
First let's look at the taxonomic distribution

```r
table(tax_table(Ps_obj)[, "Kingdom"], exclude = NULL)
```

```
## 
##  Archaea Bacteria 
##       50     9735
```

```r
table(tax_table(Ps_obj)[, "Class"], exclude = NULL)
```

```
## 
##           Abditibacteria           Acidimicrobiia           Acidobacteriae 
##                       61                      131                       17 
##           Actinobacteria      Alphaproteobacteria             Anaerolineae 
##                     1077                     1360                        8 
##            Armatimonadia                 Babeliae                  Bacilli 
##                       51                        2                      730 
##              bacteriap25              Bacteroidia BD2-11 terrestrial group 
##                        2                      985                        1 
##          Bdellovibrionia           Berkelbacteria           Blastocatellia 
##                      115                       24                       31 
##          Campylobacteria               Chlamydiae             Chloroflexia 
##                        3                        8                      477 
##               Clostridia           Coriobacteriia           Cyanobacteriia 
##                      382                       12                      474 
##          Dehalococcoidia               Deinococci       Desulfitobacteriia 
##                        2                      117                        1 
##         Desulfotomaculia         Desulfovibrionia         Desulfuromonadia 
##                        2                       12                        6 
##           Dethiobacteria            Dojkabacteria            Fibrobacteria 
##                        1                        4                        3 
##           Fimbriimonadia            Fusobacteriia      Gammaproteobacteria 
##                        4                       13                     1487 
##         Gemmatimonadetes              Gitt-GS-136          Gracilibacteria 
##                      135                        1                        3 
##             Halobacteria               Holophagae             JG30-KF-CM66 
##                       31                       15                        6 
##             Kapabacteria                   KD4-96          Kiritimatiellae 
##                        4                        9                       16 
##                Kryptonia            Lentisphaeria             Limnochordia 
##                        1                       17                        2 
##            Longimicrobia                MB-A2-108          Methanobacteria 
##                      242                        4                        7 
##           Microgenomatia               Myxococcia              Nanosalinia 
##                        2                      109                        1 
##            Negativicutes          Nitrososphaeria              Nitrospiria 
##                       26                       10                        1 
##                    OLB14              Oligoflexia                    OM190 
##                        2                       73                        1 
##                   P2-11E            Parcubacteria            Phycisphaerae 
##                        1                       32                      118 
##           Planctomycetes                Polyangia             Rhodothermia 
##                      143                      126                       10 
##            Rubrobacteria  S0134 terrestrial group          Saccharimonadia 
##                      114                        1                      234 
##        Sericytochromatia             Spirochaetia              Subgroup 22 
##                       21                        3                        2 
##               Sumerlaeia          Symbiobacteriia              Synergistia 
##                        1                        1                        1 
##     Thermoanaerobacteria          Thermoleophilia           Thermoplasmata 
##                       23                      281                        1 
##                     TK10             Unclassified                vadinHA49 
##                       22                      250                        6 
##         Vampirivibrionia         Verrucomicrobiae         Vicinamibacteria 
##                        7                       51                       15
```

```r
# table(tax_table(Ps_obj)[, "Family"], exclude = NULL)
```
Now let's remove some taxa which are obvious artefacts or those which aren't bacteria or archaea


```r
domains2remove <- c("", "Eukaryota", "Unclassified")
orders2remove <- c("Chloroplast")
families2remove <- c("Mitochondria")

Ps_obj_domains <- tax_glom(Ps_obj, "Kingdom")
Ps_obj_orders <- tax_glom(Ps_obj, "Order")
Ps_obj_families <- tax_glom(Ps_obj, "Family")

Ps_obj_tax_filt <- subset_taxa(Ps_obj, !is.na(Phylum) &
                        !Kingdom %in% domains2remove &
                      !Order %in% orders2remove &
                      !Family %in% families2remove)

sample_data(Ps_obj_tax_filt)$Library.size <- rowSums(otu_table(Ps_obj_tax_filt))
```

This removed:

```r
Summary_pruned <- tibble(
  Level = c("Kingdom", "Order", "Family"),
  ASVs.removed = c(
    table(tax_table(Ps_obj)[, "Kingdom"], exclude = NULL) %>% as.data.frame() %>% .[.$Var1 == "" | .$Var1 == "Eukaryota" | .$Var1 == "Unclassified", 2] %>% sum(),
    table(tax_table(Ps_obj)[, "Order"], exclude = NULL) %>% as.data.frame() %>% .[.$Var1 == "Chloroplast", 2] %>% sum(),
    table(tax_table(Ps_obj)[, "Family"], exclude = NULL) %>% as.data.frame() %>% .[.$Var1 == "Mitochondria", 2] %>% sum()
                     ),
  Seqs.removed = c(
    psmelt(Ps_obj_domains) %>%
      group_by(Kingdom) %>%
      filter(Kingdom == "" |
               Kingdom == "Eukaryota" | Kingdom == "Unclassified") %>%
      summarise(sum = sum(Abundance)) %>% .$sum %>% sum(),
    psmelt(Ps_obj_orders) %>%
      group_by(Order) %>%
      filter(Order == orders2remove) %>%
      summarise(sum = sum(Abundance)) %>% .$sum %>% sum(),
    psmelt(Ps_obj_families) %>%
      group_by(Family) %>%
      filter(Family == families2remove) %>%
      summarise(sum = sum(Abundance)) %>% .$sum %>% sum()
    )
  )

Summary_pruned %>% 
  kable(., digits = c(0, 1, 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

<table class="table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Level </th>
   <th style="text-align:right;"> ASVs.removed </th>
   <th style="text-align:right;"> Seqs.removed </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> Kingdom </td>
   <td style="text-align:right;"> 0 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Order </td>
   <td style="text-align:right;"> 82 </td>
   <td style="text-align:right;"> 72552 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Family </td>
   <td style="text-align:right;"> 54 </td>
   <td style="text-align:right;"> 16250 </td>
  </tr>
</tbody>
</table>

Removed 0.7398% of the sequences.

Now let's explore the prevalence of different taxa in the database.
Prevalence is the number of samples in which a taxa appears at least once. So "Mean prevalence" refers to in how many samples does a sequence belonging to the phylum appears on average, and "Sum prevalence" is the sum of all samples where any sequence from the taxon appears.

```r
prevdf <- apply(X = otu_table(Ps_obj_tax_filt),
                 MARGIN = ifelse(taxa_are_rows(Ps_obj_tax_filt), yes = 1, no = 2),
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevdf <- data.frame(Prevalence = prevdf,
                      TotalAbundance = taxa_sums(Ps_obj_tax_filt),
                      tax_table(Ps_obj_tax_filt))

prevdf %>%
  group_by(Phylum) %>%
  summarise(`Mean prevalence` = mean(Prevalence),
            `Sum prevalence` = sum(Prevalence)) ->
  Prevalence_phylum_summary

Prevalence_phylum_summary %>% 
  kable(., digits = c(0, 1, 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

<table class="table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Phylum </th>
   <th style="text-align:right;"> Mean prevalence </th>
   <th style="text-align:right;"> Sum prevalence </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> Abditibacteriota </td>
   <td style="text-align:right;"> 3.9 </td>
   <td style="text-align:right;"> 235 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Acidobacteriota </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 211 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Actinobacteriota </td>
   <td style="text-align:right;"> 3.3 </td>
   <td style="text-align:right;"> 5290 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Armatimonadota </td>
   <td style="text-align:right;"> 3.2 </td>
   <td style="text-align:right;"> 217 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bacteroidota </td>
   <td style="text-align:right;"> 2.8 </td>
   <td style="text-align:right;"> 2820 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bdellovibrionota </td>
   <td style="text-align:right;"> 2.3 </td>
   <td style="text-align:right;"> 424 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Campilobacterota </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Chloroflexi </td>
   <td style="text-align:right;"> 2.9 </td>
   <td style="text-align:right;"> 1612 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Crenarchaeota </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 30 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Cyanobacteria </td>
   <td style="text-align:right;"> 3.6 </td>
   <td style="text-align:right;"> 1501 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Deinococcota </td>
   <td style="text-align:right;"> 3.4 </td>
   <td style="text-align:right;"> 397 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Dependentiae </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Desulfobacterota </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 20 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Euryarchaeota </td>
   <td style="text-align:right;"> 1.3 </td>
   <td style="text-align:right;"> 9 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Fibrobacterota </td>
   <td style="text-align:right;"> 3.3 </td>
   <td style="text-align:right;"> 10 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Firmicutes </td>
   <td style="text-align:right;"> 1.9 </td>
   <td style="text-align:right;"> 2214 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Fusobacteriota </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 14 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gemmatimonadota </td>
   <td style="text-align:right;"> 3.4 </td>
   <td style="text-align:right;"> 1299 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Halobacterota </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 34 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> MBNT15 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Myxococcota </td>
   <td style="text-align:right;"> 3.2 </td>
   <td style="text-align:right;"> 760 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nanohaloarchaeota </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nitrospirota </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Patescibacteria </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 754 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Planctomycetota </td>
   <td style="text-align:right;"> 1.9 </td>
   <td style="text-align:right;"> 512 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Proteobacteria </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 7037 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> SAR324 clade(Marine group B) </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 7 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Spirochaetota </td>
   <td style="text-align:right;"> 1.7 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Sumerlaeota </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Synergistota </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermoplasmatota </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Unclassified </td>
   <td style="text-align:right;"> 2.1 </td>
   <td style="text-align:right;"> 255 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Verrucomicrobiota </td>
   <td style="text-align:right;"> 1.4 </td>
   <td style="text-align:right;"> 131 </td>
  </tr>
</tbody>
</table>

```r
prevdf %>%
  group_by(Order) %>%
  summarise(`Mean prevalence` = mean(Prevalence),
            `Sum prevalence` = sum(Prevalence)) ->
  Prevalence_order_summary

Prevalence_order_summary %>% 
  kable(., digits = c(0, 1, 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

<table class="table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Order </th>
   <th style="text-align:right;"> Mean prevalence </th>
   <th style="text-align:right;"> Sum prevalence </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> 0319-6G20 </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 96 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 0319-7L14 </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Abditibacteriales </td>
   <td style="text-align:right;"> 3.9 </td>
   <td style="text-align:right;"> 235 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Absconditabacteriales (SR1) </td>
   <td style="text-align:right;"> 0.0 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Acetobacterales </td>
   <td style="text-align:right;"> 2.7 </td>
   <td style="text-align:right;"> 587 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Acholeplasmatales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Acidaminococcales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Acidobacteriales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Actinomarinales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Actinomycetales </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 10 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Aeromonadales </td>
   <td style="text-align:right;"> 0.9 </td>
   <td style="text-align:right;"> 339 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Alicyclobacillales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Alteromonadales </td>
   <td style="text-align:right;"> 4.6 </td>
   <td style="text-align:right;"> 105 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Anaerolineales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Aneurinibacillales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Ardenticatenales </td>
   <td style="text-align:right;"> 1.7 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Armatimonadales </td>
   <td style="text-align:right;"> 3.5 </td>
   <td style="text-align:right;"> 177 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Azospirillales </td>
   <td style="text-align:right;"> 3.2 </td>
   <td style="text-align:right;"> 155 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Babeliales </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bacillales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 560 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bacteriovoracales </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 122 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bacteroidales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 146 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bacteroidetes VC2.1 Bac22 </td>
   <td style="text-align:right;"> 8.0 </td>
   <td style="text-align:right;"> 8 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bdellovibrionales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 136 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bifidobacteriales </td>
   <td style="text-align:right;"> 1.4 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Blastocatellales </td>
   <td style="text-align:right;"> 3.6 </td>
   <td style="text-align:right;"> 94 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Blfdi19 </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Bryobacterales </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 21 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Burkholderiales </td>
   <td style="text-align:right;"> 2.7 </td>
   <td style="text-align:right;"> 1137 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caedibacterales </td>
   <td style="text-align:right;"> 0.0 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caenarcaniphilales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caldalkalibacillales </td>
   <td style="text-align:right;"> 5.1 </td>
   <td style="text-align:right;"> 231 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caldicellulosiruptorales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caldilineales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Campylobacterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Candidatus Adlerbacteria </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Candidatus Daviesbacteria </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Candidatus Doudnabacteria </td>
   <td style="text-align:right;"> 6.0 </td>
   <td style="text-align:right;"> 6 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Candidatus Kaiserbacteria </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 7 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Candidatus Zambryskibacteria </td>
   <td style="text-align:right;"> 1.3 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Caulobacterales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 266 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> CCD24 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Cellvibrionales </td>
   <td style="text-align:right;"> 3.7 </td>
   <td style="text-align:right;"> 11 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Chitinophagales </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 259 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Chlamydiales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 9 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Chloroflexales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 80 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Christensenellales </td>
   <td style="text-align:right;"> 0.9 </td>
   <td style="text-align:right;"> 29 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Chthoniobacterales </td>
   <td style="text-align:right;"> 2.2 </td>
   <td style="text-align:right;"> 69 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Clostridia UCG-014 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 6 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Clostridia vadinBB60 group </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 14 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Clostridiales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 17 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Coriobacteriales </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 14 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Corynebacteriales </td>
   <td style="text-align:right;"> 2.3 </td>
   <td style="text-align:right;"> 288 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Coxiellales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 6 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Cyanobacteriales </td>
   <td style="text-align:right;"> 3.6 </td>
   <td style="text-align:right;"> 1053 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Cytophagales </td>
   <td style="text-align:right;"> 3.6 </td>
   <td style="text-align:right;"> 1855 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Deinococcales </td>
   <td style="text-align:right;"> 3.5 </td>
   <td style="text-align:right;"> 373 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Desulfitobacteriales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Desulfotomaculales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Desulfovibrionales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 12 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Dethiobacterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Diplorickettsiales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 32 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Elsterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Enterobacterales </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 106 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Entomoplasmatales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Erysipelotrichales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 27 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Eubacteriales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Euzebyales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 23 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> EV818SWSAP88 </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Exiguobacterales </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Ferrovibrionales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Fibrobacterales </td>
   <td style="text-align:right;"> 3.3 </td>
   <td style="text-align:right;"> 10 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Fimbriimonadales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 7 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Flavobacteriales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 244 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Frankiales </td>
   <td style="text-align:right;"> 4.6 </td>
   <td style="text-align:right;"> 1482 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Fusobacteriales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 14 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gaiellales </td>
   <td style="text-align:right;"> 2.2 </td>
   <td style="text-align:right;"> 50 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gammaproteobacteria Incertae Sedis </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gastranaerophilales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gemmatales </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 7 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Gemmatimonadales </td>
   <td style="text-align:right;"> 3.2 </td>
   <td style="text-align:right;"> 427 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Geoalkalibacteraceae </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Geobacterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Glycomycetales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Haliangiales </td>
   <td style="text-align:right;"> 3.4 </td>
   <td style="text-align:right;"> 126 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Halobacterales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 34 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Hungateiclostridiaceae </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> IMCC26256 </td>
   <td style="text-align:right;"> 1.4 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Isosphaerales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 204 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Izemoplasmatales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 6 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> JGI 0000069-P22 </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Kallotenuales </td>
   <td style="text-align:right;"> 3.3 </td>
   <td style="text-align:right;"> 1003 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Kapabacteriales </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 6 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Kineosporiales </td>
   <td style="text-align:right;"> 4.7 </td>
   <td style="text-align:right;"> 128 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Kryptoniales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Lachnospirales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 60 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Lactobacillales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 320 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Legionellales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Leptolyngbyales </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 7 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Limnochordales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Longimicrobiales </td>
   <td style="text-align:right;"> 3.5 </td>
   <td style="text-align:right;"> 857 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Methanobacteriales </td>
   <td style="text-align:right;"> 1.3 </td>
   <td style="text-align:right;"> 9 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Methanomassiliicoccales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Methylacidiphilales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Micrococcales </td>
   <td style="text-align:right;"> 2.5 </td>
   <td style="text-align:right;"> 692 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Micromonosporales </td>
   <td style="text-align:right;"> 4.0 </td>
   <td style="text-align:right;"> 163 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Microtrichales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 123 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> mle1-27 </td>
   <td style="text-align:right;"> 3.8 </td>
   <td style="text-align:right;"> 73 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Monoglobales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> MSB-4B10 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Mycoplasmatales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Myxococcales </td>
   <td style="text-align:right;"> 3.7 </td>
   <td style="text-align:right;"> 403 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nannocystales </td>
   <td style="text-align:right;"> 12.0 </td>
   <td style="text-align:right;"> 36 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nanosalinales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nitrosococcales </td>
   <td style="text-align:right;"> 4.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nitrososphaerales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 30 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Nitrospirales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Obscuribacterales </td>
   <td style="text-align:right;"> 0.0 </td>
   <td style="text-align:right;"> 0 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Oceanospirillales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 90 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Oligoflexales </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Oligosphaerales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Opitutales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Oscillospirales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 159 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Oxyphotobacteria Incertae Sedis </td>
   <td style="text-align:right;"> 3.7 </td>
   <td style="text-align:right;"> 304 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Paenibacillales </td>
   <td style="text-align:right;"> 2.7 </td>
   <td style="text-align:right;"> 261 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Paracaedibacterales </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Parvibaculales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pasteurellales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 23 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pedosphaerales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Peptococcales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Peptostreptococcales-Tissierellales </td>
   <td style="text-align:right;"> 1.3 </td>
   <td style="text-align:right;"> 79 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Phycisphaerales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pirellulales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 19 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Planctomycetales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 9 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> PLTA13 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Polyangiales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 110 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Propionibacteriales </td>
   <td style="text-align:right;"> 4.1 </td>
   <td style="text-align:right;"> 652 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pseudomonadales </td>
   <td style="text-align:right;"> 2.7 </td>
   <td style="text-align:right;"> 987 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pseudonocardiales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 122 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Puniceispirillales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Pyrinomonadales </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> R7C24 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Reyranellales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> RF39 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 13 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rhizobiales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 1094 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rhodobacterales </td>
   <td style="text-align:right;"> 3.9 </td>
   <td style="text-align:right;"> 595 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rhodospirillales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 8 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rhodothermales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 18 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rickettsiales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Rubrobacterales </td>
   <td style="text-align:right;"> 4.4 </td>
   <td style="text-align:right;"> 501 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> S085 </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Saccharimonadales </td>
   <td style="text-align:right;"> 2.7 </td>
   <td style="text-align:right;"> 633 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Salinisphaerales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> SAR11 clade </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> SBR1031 </td>
   <td style="text-align:right;"> 1.5 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Silvanigrellales </td>
   <td style="text-align:right;"> 2.9 </td>
   <td style="text-align:right;"> 57 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Solibacterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 4 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Solirubrobacterales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 751 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Sphingobacteriales </td>
   <td style="text-align:right;"> 3.2 </td>
   <td style="text-align:right;"> 280 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Sphingomonadales </td>
   <td style="text-align:right;"> 3.8 </td>
   <td style="text-align:right;"> 1094 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Spirochaetales </td>
   <td style="text-align:right;"> 1.7 </td>
   <td style="text-align:right;"> 5 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Staphylococcales </td>
   <td style="text-align:right;"> 2.6 </td>
   <td style="text-align:right;"> 246 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Steroidobacterales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Streptomycetales </td>
   <td style="text-align:right;"> 1.3 </td>
   <td style="text-align:right;"> 18 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Streptosporangiales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 48 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Subgroup 13 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Subgroup 17 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Subgroup 2 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Subgroup 7 </td>
   <td style="text-align:right;"> 3.7 </td>
   <td style="text-align:right;"> 56 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Sumerlaeales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Symbiobacteriales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Synergistales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 1 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Tepidisphaerales </td>
   <td style="text-align:right;"> 2.1 </td>
   <td style="text-align:right;"> 247 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermales </td>
   <td style="text-align:right;"> 2.4 </td>
   <td style="text-align:right;"> 22 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermicanales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 61 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermoactinomycetales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermoanaerobacterales </td>
   <td style="text-align:right;"> 1.6 </td>
   <td style="text-align:right;"> 18 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermobaculales </td>
   <td style="text-align:right;"> 3.4 </td>
   <td style="text-align:right;"> 44 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermomicrobiales </td>
   <td style="text-align:right;"> 3.0 </td>
   <td style="text-align:right;"> 355 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Thermosynechococcales </td>
   <td style="text-align:right;"> 4.5 </td>
   <td style="text-align:right;"> 59 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Tistrellales </td>
   <td style="text-align:right;"> 2.0 </td>
   <td style="text-align:right;"> 40 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Unclassified </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 971 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Vampirovibrionales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 2 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Veillonellales-Selenomonadales </td>
   <td style="text-align:right;"> 1.2 </td>
   <td style="text-align:right;"> 29 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Verrucomicrobiales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 9 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Vibrionales </td>
   <td style="text-align:right;"> 1.9 </td>
   <td style="text-align:right;"> 15 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Vicinamibacterales </td>
   <td style="text-align:right;"> 1.1 </td>
   <td style="text-align:right;"> 15 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Victivallales </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 16 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> WCHB1-41 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 16 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> WD260 </td>
   <td style="text-align:right;"> 1.0 </td>
   <td style="text-align:right;"> 3 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Xanthomonadales </td>
   <td style="text-align:right;"> 1.8 </td>
   <td style="text-align:right;"> 181 </td>
  </tr>
</tbody>
</table>

Based on that I'll remove all orders with a sum prevalence of under 5% (10.1) of all samples

```r
Prevalence_order_summary %>% 
  filter(`Sum prevalence` < (0.05 * nsamples(Ps_obj_tax_filt))) %>% 
  dplyr::select(Order) %>% 
  map(as.character) %>% 
  unlist() ->
  filterOrder

Ps_obj_order_prev_filt <- subset_taxa(Ps_obj_tax_filt, !Order %in% filterOrder)
# Taxonomy.bs %<>% 
#   filter(Taxonomy.bs$Name %in% row.names(Ps_obj_order_prev_filt@tax_table))
sample_data(Ps_obj_order_prev_filt)$Library.size <- rowSums(otu_table(Ps_obj_order_prev_filt))
print(Ps_obj_tax_filt)
```

```
## phyloseq-class experiment-level object
## otu_table()   OTU Table:          [ 9649 taxa and 202 samples ]:
## sample_data() Sample Data:        [ 202 samples by 26 sample variables ]:
## tax_table()   Taxonomy Table:     [ 9649 taxa by 6 taxonomic ranks ]:
## refseq()      DNAStringSet:       [ 9649 reference sequences ]
## taxa are columns
```

```r
print(Ps_obj_order_prev_filt)
```

```
## phyloseq-class experiment-level object
## otu_table()   OTU Table:          [ 9416 taxa and 202 samples ]:
## sample_data() Sample Data:        [ 202 samples by 26 sample variables ]:
## tax_table()   Taxonomy Table:     [ 9416 taxa by 6 taxonomic ranks ]:
## refseq()      DNAStringSet:       [ 9416 reference sequences ]
## taxa are columns
```

This removed 233 or 2% of the ASVs, and 0.597% of the reads.

Plot general prevalence features of the phyla

```r
# Subset to the remaining phyla
prevdf_phylum_filt <- subset(prevdf, Phylum %in% get_taxa_unique(Ps_obj_order_prev_filt, "Phylum"))
ggplot(prevdf_phylum_filt,
       aes(TotalAbundance, Prevalence / nsamples(Ps_obj_order_prev_filt), color = Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.05,
             alpha = 0.5,
             linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap( ~ Phylum) + theme(legend.position = "none")
```

![](03_Taxonomy_figures/prevalence phylum-1.svg)<!-- -->

Plot general prevalence features of the top 20 orders

```r
# Subset to the remaining phyla
prevdf_order_filt <- subset(prevdf, Order %in% get_taxa_unique(Ps_obj_order_prev_filt, "Order"))

# grab the top 30 most abundant orders
prevdf_order_filt %>% 
  group_by(Order) %>%
  summarise(Combined.abundance = sum(TotalAbundance)) %>% 
  arrange(desc(Combined.abundance)) %>% 
  .[1:30, "Order"]  ->
  Orders2plot

prevdf_order_filt2 <- subset(prevdf, Order %in% Orders2plot$Order)

ggplot(prevdf_order_filt2,
       aes(TotalAbundance, Prevalence / nsamples(Ps_obj_order_prev_filt), color = Order)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.05,
             alpha = 0.5,
             linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap( ~ Order) + theme(legend.position = "none")
```

![](03_Taxonomy_figures/prevalence order-1.svg)<!-- -->

#### Unsupervised filtering by prevalence
I'll remove all sequences which appear in less than 5% of the samples

```r
# Define prevalence threshold as 5% of total samples
prevalenceThreshold <- 0.05 * nsamples(Ps_obj_order_prev_filt)
prevalenceThreshold
```

```
## [1] 10.1
```

```r
# Execute prevalence filter, using `prune_taxa()` function
keepTaxa <-
  row.names(prevdf_phylum_filt)[(prevdf_phylum_filt$Prevalence >= prevalenceThreshold)]
Ps_obj_seq_prev_filt <- prune_taxa(keepTaxa, Ps_obj_order_prev_filt)
# Taxonomy.bs %<>% 
#   filter(Taxonomy.bs$Name %in% row.names(Ps_obj_seq_prev_filt@tax_table))
sample_data(Ps_obj_seq_prev_filt)$Library.size <- rowSums(otu_table(Ps_obj_seq_prev_filt))
print(Ps_obj_order_prev_filt)
```

```
## phyloseq-class experiment-level object
## otu_table()   OTU Table:          [ 9416 taxa and 202 samples ]:
## sample_data() Sample Data:        [ 202 samples by 26 sample variables ]:
## tax_table()   Taxonomy Table:     [ 9416 taxa by 6 taxonomic ranks ]:
## refseq()      DNAStringSet:       [ 9416 reference sequences ]
## taxa are columns
```

```r
print(Ps_obj_seq_prev_filt)
```

```
## phyloseq-class experiment-level object
## otu_table()   OTU Table:          [ 471 taxa and 202 samples ]:
## sample_data() Sample Data:        [ 202 samples by 26 sample variables ]:
## tax_table()   Taxonomy Table:     [ 471 taxa by 6 taxonomic ranks ]:
## refseq()      DNAStringSet:       [ 471 reference sequences ]
## taxa are columns
```
This removed 8945 or 95% of the ASVs! But only 27.915% of the reads.

However all these removed ASVs accounted for only: 

```r
prevdf_phylum_filt %>% 
  arrange(., Prevalence) %>% 
  group_by(Prevalence > prevalenceThreshold) %>% 
  summarise(Abundance = sum(TotalAbundance)) %>%
  mutate(`Rel. Ab.` = percent(Abundance / sum(Abundance), accuracy = 0.01)) %>% 
  kable(., digits = c(0, 1, 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

<table class="table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Prevalence &gt; prevalenceThreshold </th>
   <th style="text-align:right;"> Abundance </th>
   <th style="text-align:left;"> Rel. Ab. </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> FALSE </td>
   <td style="text-align:right;"> 2745595 </td>
   <td style="text-align:left;"> 28.28% </td>
  </tr>
  <tr>
   <td style="text-align:left;"> TRUE </td>
   <td style="text-align:right;"> 6963743 </td>
   <td style="text-align:left;"> 71.72% </td>
  </tr>
</tbody>
</table>
So it's fine to remove them.

### General taxonomic features

```r
Ps_obj_seq_prev_filt %>%
  subset_samples(., sample_sums(Ps_obj_seq_prev_filt) > min_lib_size) %>%
  filter_taxa(., function(x)
    sum(x) > 0, TRUE) %>%
  transform_sample_counts(., function(x)
    x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset

Ps_obj_seq_prev_filt_subset %>% 
  sample_data() %>% 
  as_tibble() %>% 
  arrange(Sample) %>% 
  pull(.sample) ->
  Sample.order

Ps_obj_seq_prev_filt_subset_100 <-
  prune_taxa(names(sort(taxa_sums(Ps_obj_seq_prev_filt_subset), TRUE)[1:100]), Ps_obj_seq_prev_filt_subset)
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_100,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  taxa.label = "Order",
  sample.order = Sample.order,
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/seqs heatmaps-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset_firmi <-
  subset_taxa(Ps_obj_seq_prev_filt_subset, Phylum == "Firmicutes")
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_firmi,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/firmicutes heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>%
  subset_taxa(., Phylum == "Actinobacteriota") %>%
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_actino
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_actino,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Actinobacteria heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>%
  subset_taxa(., Phylum == "Bacteroidota") %>%
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_bacter

plot_heatmap(
  Ps_obj_seq_prev_filt_subset_bacter,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Bacteroidetes heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>%
  subset_taxa(., Phylum == "Proteobacteria") %>%
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_proteo

plot_heatmap(
  Ps_obj_seq_prev_filt_subset_proteo,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Phylum",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Proteobacteria heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>% 
  subset_taxa(., Order == "Enterobacterales") %>% 
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_chitino
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_chitino,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Enterobacterales heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>% 
  subset_taxa(., Order == "Chitinophagales") %>% 
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_chitino
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_chitino,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Chitinophagales heatmap-1.svg)<!-- -->


```r
Ps_obj_seq_prev_filt_subset %>% 
  subset_taxa(., Order == "Rhizobiales") %>% 
  transform_sample_counts(., function(x) x / sum(x) * 100) ->
  Ps_obj_seq_prev_filt_subset_rhizo
plot_heatmap(
  Ps_obj_seq_prev_filt_subset_rhizo,
  method = NULL,
  distance = NULL,
  sample.label = "Sample",
  sample.order = Sample.order,
  taxa.label = "Order",
  low = "#000033",
  high = "#FF3300"
)
```

![](03_Taxonomy_figures/Rhizobiales heatmap-1.svg)<!-- -->


```r
# Ps_obj_seq_prev_filt_subset %>% 
#   subset_taxa(., Order == "Betaproteobacteriales") %>% 
#   transform_sample_counts(., function(x) x / sum(x) * 100) ->
#   Ps_obj_seq_prev_filt_subset_beta
# plot_heatmap(
#   Ps_obj_seq_prev_filt_subset_beta,
#   method = NULL,
#   distance = NULL,
#   sample.label = "Sample.name",
#   sample.order = Sample.order,
#   taxa.label = "Order",
#   low = "#000033",
#   high = "#FF3300"
# )
```

Let us look at the agglomerated taxa

```r
Ps_obj_seq_prev_filt_subset %>% 
  transform_sample_counts(., function(x) x / sum(x) * 100) %>% 
  tax_glom(., "Phylum", NArm = TRUE) ->
  Ps_obj_seq_prev_glom

plot_heatmap(
  Ps_obj_seq_prev_glom,
  # method = "NMDS",
  # distance = "bray",
  sample.order = Sample.order,
  sample.label = "Sample",
  taxa.label = "Phylum",
  taxa.order = "Phylum",
  low = "#000033",
  high = "#FF3300"
) + theme_bw(base_size = 20) + theme(axis.text.x = element_text(hjust = 1.0, angle = 45.0))
```

![](03_Taxonomy_figures/agglomerated taxa heatmap-1.svg)<!-- -->


#### Explore abundance distribution of specific taxa

```r
Ps_obj_seq_prev_filt_subset_ra <- transform_sample_counts(Ps_obj_seq_prev_filt_subset, function(x){x / sum(x)} * 100)
plot_abundance(Ps_obj_seq_prev_filt_subset_ra, taxa = "Proteobacteria")
```

![](03_Taxonomy_figures/abund dist-1.svg)<!-- -->

```r
# plotBefore <- plot_abundance(Ps_obj_seq_prev_filt_subset, taxa = "Proteobacteria")
# plotAfter <- plot_abundance(Ps_obj_seq_prev_filt_subset_ra, taxa = "Proteobacteria")
# Combine each plot into one graphic.
# grid.arrange(nrow = 2, plotBefore, plotAfter)
```


```r
Ps_obj_seq_prev_filt_subset_ra_taxa <- subset_taxa(Ps_obj_seq_prev_filt_subset_ra, Order == "Bacillales")
plot_abundance(Ps_obj_seq_prev_filt_subset_ra_taxa, Facet = "Genus")
```

![](03_Taxonomy_figures/specific taxa-1.svg)<!-- -->

### Taxa violin plots

```r
taxa_order <- order_taxa(Ps_obj_seq_prev_filt_subset)
Ps_obj_seq_prev_filt_subset_grouped <- mark_rare_taxa(Ps_obj_seq_prev_filt_subset, rank = "Phylum", rare_thresh = 0.01)
plot_tax_violin(Ps_obj_seq_prev_filt_subset_grouped, grouping_var1 = Var3, grouping_var2 = Var2, taxa_order = taxa_order)
```

![](03_Taxonomy_figures/violin plots-1.svg)<!-- -->

#### Make Krona plots

```r
dir.create(paste0(data_path, "Krona/")) # fore the tables
for (i in seq(nsamples(Ps_obj_seq_prev_filt_subset))) {
  sample_data <-
    data.frame(t(otu_table(Ps_obj_seq_prev_filt_subset)[i,]), tax_table(Ps_obj_seq_prev_filt_subset))
  sample_data <- sample_data[sample_data[, 1] > 0, ]
  write_tsv(sample_data, paste0(data_path, "Krona/", sample_names(Ps_obj_seq_prev_filt_subset)[i], ".tsv"))
}

sample_data(Ps_obj_seq_prev_filt_subset_ra)$Krona_combinations <- paste0(
  get_variable(Ps_obj_seq_prev_filt_subset_ra, Var3),
  ".",
  get_variable(Ps_obj_seq_prev_filt_subset_ra, Var2),
  ".",
  get_variable(Ps_obj_seq_prev_filt_subset_ra, Var3)
)

Ps_obj_seq_prev_filt_subset_ra %>%
  phyloseq::merge_samples(., "Krona_combinations", fun = mean) ->
  merged_Ps_obj

for (i in seq(nsamples(merged_Ps_obj))) {
  sample_data <-
    data.frame(t(otu_table(merged_Ps_obj)[i,]), tax_table(merged_Ps_obj))
  sample_data <- sample_data[sample_data[, 1] > 0, ]
  write_tsv(sample_data, paste0(data_path, "Krona/", Proj_name, ".", sample_names(merged_Ps_obj)[i], ".tsv"))
}

list.files(paste0(data_path, "Krona/"), full.names = TRUE) %>%
  paste(., collapse = " ") %>%
  paste0("/usr/local/bin/ktImportText ", .,
         " -o ",
         Proj_name,
         "_Krona.html") %>%
  system()
```

#### Save filtered phyloseq object

```r
saveRDS(Ps_obj_tax_filt, file = paste0(data_path, Proj_name, "_tax_filt.Rds"))
# save seqs
Ps_obj_tax_filt %>%  
  refseq() %>% 
  writeXStringSet(., filepath = paste0(data_path, "DADA2_seqs_tax_filt.fa"), format = "fasta", width = 1000)
saveRDS(Ps_obj_order_prev_filt, file = paste0(data_path, Proj_name, "_tax_prev_filt.Rds"))
Ps_obj_order_prev_filt %>%  
  refseq() %>% 
  writeXStringSet(., filepath = paste0(data_path, "DADA2_seqs_tax_prev_filt.fa"), format = "fasta", width = 1000)
saveRDS(Ps_obj_seq_prev_filt, file = paste0(data_path, Proj_name, "_seq_prev_filt.Rds"))
Ps_obj_seq_prev_filt %>%  
  refseq() %>% 
  writeXStringSet(., filepath = paste0(data_path, "DADA2_seqs_seq_prev_filt.fa"), format = "fasta", width = 1000)
```


```r
sessioninfo::session_info() %>%
  details::details(
    summary = 'Current session info',
    open    = TRUE
  )
```

<details open>
<summary> <span title='Click to Expand'> Current session info </span> </summary>

```r

 Session info 
 setting  value                       
 version  R version 4.1.1 (2021-08-10)
 os       Ubuntu 18.04.6 LTS          
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Prague               
 date     2021-10-14                  

 Packages 
 package          * version    date       lib source                           
 ade4               1.7-18     2021-09-16 [1] CRAN (R 4.1.1)                   
 ape                5.5        2021-04-25 [1] CRAN (R 4.1.1)                   
 assertthat         0.2.1      2019-03-21 [1] CRAN (R 4.1.1)                   
 backports          1.2.1      2020-12-09 [1] CRAN (R 4.1.1)                   
 Biobase            2.52.0     2021-05-19 [1] Bioconductor                     
 BiocGenerics     * 0.38.0     2021-05-19 [1] Bioconductor                     
 biomformat         1.20.0     2021-05-19 [1] Bioconductor                     
 Biostrings       * 2.60.2     2021-08-05 [1] Bioconductor                     
 bit                4.0.4      2020-08-04 [1] CRAN (R 4.1.1)                   
 bit64              4.0.5      2020-08-30 [1] CRAN (R 4.1.1)                   
 bitops             1.0-7      2021-04-24 [1] CRAN (R 4.1.1)                   
 broom              0.7.9      2021-07-27 [1] CRAN (R 4.1.1)                   
 cellranger         1.1.0      2016-07-27 [1] CRAN (R 4.1.1)                   
 cli                3.0.1      2021-07-17 [1] CRAN (R 4.1.1)                   
 clipr              0.7.1      2020-10-08 [1] CRAN (R 4.1.1)                   
 cluster            2.1.2      2021-04-17 [1] CRAN (R 4.1.1)                   
 codetools          0.2-18     2020-11-04 [1] CRAN (R 4.1.1)                   
 colorspace         2.0-2      2021-06-24 [1] CRAN (R 4.1.1)                   
 cowplot          * 1.1.1      2020-12-30 [1] CRAN (R 4.1.1)                   
 crayon             1.4.1      2021-02-08 [1] CRAN (R 4.1.1)                   
 data.table         1.14.2     2021-09-27 [1] CRAN (R 4.1.1)                   
 DBI                1.1.1      2021-01-15 [1] CRAN (R 4.1.1)                   
 dbplyr             2.1.1      2021-04-06 [1] CRAN (R 4.1.1)                   
 desc               1.4.0      2021-09-28 [1] CRAN (R 4.1.1)                   
 details            0.2.1      2020-01-12 [1] CRAN (R 4.1.1)                   
 digest             0.6.28     2021-09-23 [1] CRAN (R 4.1.1)                   
 dplyr            * 1.0.7      2021-06-18 [1] CRAN (R 4.1.1)                   
 ellipsis           0.3.2      2021-04-29 [1] CRAN (R 4.1.1)                   
 evaluate           0.14       2019-05-28 [1] CRAN (R 4.1.1)                   
 extrafont        * 0.17       2014-12-08 [1] CRAN (R 4.1.1)                   
 extrafontdb        1.0        2012-06-11 [1] CRAN (R 4.1.1)                   
 fansi              0.5.0      2021-05-25 [1] CRAN (R 4.1.1)                   
 farver             2.1.0      2021-02-28 [1] CRAN (R 4.1.1)                   
 fastmap            1.1.0      2021-01-25 [1] CRAN (R 4.1.1)                   
 forcats          * 0.5.1      2021-01-27 [1] CRAN (R 4.1.1)                   
 foreach            1.5.1      2020-10-15 [1] CRAN (R 4.1.1)                   
 fs                 1.5.0      2020-07-31 [1] CRAN (R 4.1.1)                   
 generics           0.1.0      2020-10-31 [1] CRAN (R 4.1.1)                   
 GenomeInfoDb     * 1.28.4     2021-09-05 [1] Bioconductor                     
 GenomeInfoDbData   1.2.6      2021-10-13 [1] Bioconductor                     
 ggplot2          * 3.3.5      2021-06-25 [1] CRAN (R 4.1.1)                   
 glue               1.4.2      2020-08-27 [1] CRAN (R 4.1.1)                   
 gtable             0.3.0      2019-03-25 [1] CRAN (R 4.1.1)                   
 haven              2.4.3      2021-08-04 [1] CRAN (R 4.1.1)                   
 highr              0.9        2021-04-16 [1] CRAN (R 4.1.1)                   
 hms                1.1.1      2021-09-26 [1] CRAN (R 4.1.1)                   
 htmltools          0.5.2      2021-08-25 [1] CRAN (R 4.1.1)                   
 httr               1.4.2      2020-07-20 [1] CRAN (R 4.1.1)                   
 igraph             1.2.6      2020-10-06 [1] CRAN (R 4.1.1)                   
 IRanges          * 2.26.0     2021-05-19 [1] Bioconductor                     
 iterators          1.0.13     2020-10-15 [1] CRAN (R 4.1.1)                   
 jquerylib          0.1.4      2021-04-26 [1] CRAN (R 4.1.1)                   
 jsonlite           1.7.2      2020-12-09 [1] CRAN (R 4.1.1)                   
 kableExtra       * 1.3.4      2021-02-20 [1] CRAN (R 4.1.1)                   
 knitr              1.36       2021-09-29 [1] CRAN (R 4.1.1)                   
 labeling           0.4.2      2020-10-20 [1] CRAN (R 4.1.1)                   
 lattice          * 0.20-45    2021-09-22 [1] CRAN (R 4.1.1)                   
 lifecycle          1.0.1      2021-09-24 [1] CRAN (R 4.1.1)                   
 lubridate          1.8.0      2021-10-07 [1] CRAN (R 4.1.1)                   
 magrittr         * 2.0.1      2020-11-17 [1] CRAN (R 4.1.1)                   
 MASS               7.3-54     2021-05-03 [1] CRAN (R 4.1.1)                   
 Matrix             1.3-4      2021-06-01 [1] CRAN (R 4.1.1)                   
 mgcv               1.8-38     2021-10-06 [1] CRAN (R 4.1.1)                   
 modelr             0.1.8      2020-05-19 [1] CRAN (R 4.1.1)                   
 multtest           2.48.0     2021-05-19 [1] Bioconductor                     
 munsell            0.5.0      2018-06-12 [1] CRAN (R 4.1.1)                   
 nlme               3.1-153    2021-09-07 [1] CRAN (R 4.1.1)                   
 permute          * 0.9-5      2019-03-12 [1] CRAN (R 4.1.1)                   
 phyloseq         * 1.36.0     2021-05-19 [1] Bioconductor                     
 pillar             1.6.3      2021-09-26 [1] CRAN (R 4.1.1)                   
 pkgconfig          2.0.3      2019-09-22 [1] CRAN (R 4.1.1)                   
 plyr               1.8.6      2020-03-03 [1] CRAN (R 4.1.1)                   
 png                0.1-7      2013-12-03 [1] CRAN (R 4.1.1)                   
 purrr            * 0.3.4      2020-04-17 [1] CRAN (R 4.1.1)                   
 R6                 2.5.1      2021-08-19 [1] CRAN (R 4.1.1)                   
 ragg               1.1.3      2021-06-09 [1] CRAN (R 4.1.1)                   
 Rcpp               1.0.7      2021-07-07 [1] CRAN (R 4.1.1)                   
 RCurl              1.98-1.5   2021-09-17 [1] CRAN (R 4.1.1)                   
 readr            * 2.0.2      2021-09-27 [1] CRAN (R 4.1.1)                   
 readxl             1.3.1      2019-03-13 [1] CRAN (R 4.1.1)                   
 reprex             2.0.1      2021-08-05 [1] CRAN (R 4.1.1)                   
 reshape2           1.4.4      2020-04-09 [1] CRAN (R 4.1.1)                   
 rhdf5              2.36.0     2021-05-19 [1] Bioconductor                     
 rhdf5filters       1.4.0      2021-05-19 [1] Bioconductor                     
 Rhdf5lib           1.14.2     2021-07-06 [1] Bioconductor                     
 rlang              0.4.11     2021-04-30 [1] CRAN (R 4.1.1)                   
 rmarkdown          2.11       2021-09-14 [1] CRAN (R 4.1.1)                   
 rprojroot          2.0.2      2020-11-15 [1] CRAN (R 4.1.1)                   
 rstudioapi         0.13       2020-11-12 [1] CRAN (R 4.1.1)                   
 Rttf2pt1           1.3.9      2021-07-22 [1] CRAN (R 4.1.1)                   
 rvest              1.0.1      2021-07-26 [1] CRAN (R 4.1.1)                   
 S4Vectors        * 0.30.2     2021-10-03 [1] Bioconductor                     
 scales           * 1.1.1      2020-05-11 [1] CRAN (R 4.1.1)                   
 see              * 0.6.8      2021-10-03 [1] CRAN (R 4.1.1)                   
 sessioninfo        1.1.1      2018-11-05 [1] CRAN (R 4.1.1)                   
 speedyseq        * 0.5.3.9018 2021-10-13 [1] Github (mikemc/speedyseq@ceb941f)
 stringi            1.7.5      2021-10-04 [1] CRAN (R 4.1.1)                   
 stringr          * 1.4.0      2019-02-10 [1] CRAN (R 4.1.1)                   
 survival           3.2-13     2021-08-24 [1] CRAN (R 4.1.1)                   
 svglite          * 2.0.0      2021-02-20 [1] CRAN (R 4.1.1)                   
 systemfonts        1.0.3      2021-10-13 [1] CRAN (R 4.1.1)                   
 textshaping        0.3.6      2021-10-13 [1] CRAN (R 4.1.1)                   
 tibble           * 3.1.5      2021-09-30 [1] CRAN (R 4.1.1)                   
 tidyr            * 1.1.4      2021-09-27 [1] CRAN (R 4.1.1)                   
 tidyselect         1.1.1      2021-04-30 [1] CRAN (R 4.1.1)                   
 tidyverse        * 1.3.1      2021-04-15 [1] CRAN (R 4.1.1)                   
 tzdb               0.1.2      2021-07-20 [1] CRAN (R 4.1.1)                   
 utf8               1.2.2      2021-07-24 [1] CRAN (R 4.1.1)                   
 vctrs              0.3.8      2021-04-29 [1] CRAN (R 4.1.1)                   
 vegan            * 2.5-7      2020-11-28 [1] CRAN (R 4.1.1)                   
 viridisLite        0.4.0      2021-04-13 [1] CRAN (R 4.1.1)                   
 vroom              1.5.5      2021-09-14 [1] CRAN (R 4.1.1)                   
 webshot            0.5.2      2019-11-22 [1] CRAN (R 4.1.1)                   
 withr              2.4.2      2021-04-18 [1] CRAN (R 4.1.1)                   
 xfun               0.26       2021-09-14 [1] CRAN (R 4.1.1)                   
 xml2               1.3.2      2020-04-23 [1] CRAN (R 4.1.1)                   
 XVector          * 0.32.0     2021-05-19 [1] Bioconductor                     
 yaml               2.2.1      2020-02-01 [1] CRAN (R 4.1.1)                   
 zlibbioc           1.38.0     2021-05-19 [1] Bioconductor                     

[1] /home/gabay/R/x86_64-pc-linux-gnu-library/4.1
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library

```

</details>
<br>

```r
## References
```
