---
title: "Zin restoration SIP"
subtitle: "04 Read-depth normalisation attempts for DADA2-based analysis"
description: "V1.1"
author: "Roey Angel"
date: "2021-10-21"
bibliography: references.bib
link-citations: yes
csl: fems-microbiology-ecology.csl
output:
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    keep_md: true
    number_sections: false
    highlight: "pygments"
    theme: "flatly"
    dev: "png"
    df_print: "kable"
    fig_caption: true
    code_folding: "show"
---






[roey.angel@bc.cas.cz](mailto: roey.angel@bc.cas.cz)  

## Read-depth distribution and normalisation
This analysis tests the effect of library read-depth distribution on the community composition. It then performs various read-depth normalisation methods on the DADA2 zOTU-based dataset, for determining the optimal strategy to handle the bias of uneven read distribution. 

### Setting general parameters:

```r
set.seed(1000)
min_lib_size <- 2500
data_path <- "./DADA2_pseudo/"
# Metadata_table <- "./Preservation_methods_metadataJS.csv"
# Seq_table <- "DADA2.seqtab_nochim_decontam.tsv"
Proj_name <- "Zin_SIP"
Ps_file <- paste0(Proj_name, "_seq_prev_filt.Rds")
Var1 = "Density..g.ml.1." # e.g sampling point / replicate
Var2 = "Treatment" # e.g. a treatment or a manipulation
Var3 = "Label..18O." # e.g. a treatment/manipulation or an important covariant
Var4 = "" # e.g. an important covariant

adonis_f <- as.formula(
  paste("test_mat",
    paste(c(Var3, Var2), collapse = " * "),
  sep = " ~ "))
```

### Load phyloseq object
Read a decontaminated and filtered phyloseq object generated by [03_Taxonomical_analysis.Rmd](03_Taxonomical_analysis.Rmd). 


```r
readRDS(paste0(data_path, Ps_file)) ->
  Ps_obj
```

### Exploring dataset features
First let us look at the count data distribution

```r
qplot(rowSums(otu_table(Ps_obj)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/plot abundance-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/plot abundance-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj, Var1, Var2, Var3)
```

![](04_Normalisation_figures/plot abundance-3.svg)<!-- -->

```r
sample_data(Ps_obj) %>% 
  as_tibble() %>% 
  select(Read1_file, Library.size) %>% 
  as(., "data.frame") %>% 
  kable(.) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

<table class="table table-striped table-hover table-condensed table-responsive" style="width: auto !important; margin-left: auto; margin-right: auto;">
 <thead>
  <tr>
   <th style="text-align:left;"> Read1_file </th>
   <th style="text-align:right;"> Library.size </th>
  </tr>
 </thead>
<tbody>
  <tr>
   <td style="text-align:left;"> 661Roey001-N1-2_S1_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8289 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey002-N1-3_S2_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 20206 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey003-N1-4_S3_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17596 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey004-N1-5_S4_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21856 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey005-N1-6_S5_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 22652 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey006-N1-7_S6_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 87691 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey007-N1-8_S7_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 180042 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey008-N1-9_S8_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 105478 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey009-N1-10_S9_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 134196 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey010-N1-11_S10_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 62139 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey012-N3-2_S12_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 81442 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey013-N3-3_S13_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 48766 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey014-N3-4_S14_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 26202 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey015-N3-5_S15_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 22280 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey016-N3-6_S16_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 24479 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey017-N3-7_S17_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 49054 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey018-N3-8_S18_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 213053 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey019-N3-9_S19_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 222973 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey020-N3-10_S20_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 173218 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey021-N3-11_S21_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 197036 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey022-N5-2_S22_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4862 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey023-N5-3_S23_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9579 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey024-N5-4_S24_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10905 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey025-N5-5_S25_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10047 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey026-N5-6_S26_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9339 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey027-N5-7_S27_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14576 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey028-N5-8_S28_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10942 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey029-N5-9_S29_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12347 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey030-N5-10_S30_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14612 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey031-N5-11_S31_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 7592 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey032-N7-2_S32_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18402 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey033-N7-3_S33_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12430 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey034-N7-4_S34_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15754 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey035-N7-5_S35_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12532 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey036-N7-6_S36_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16187 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey037-N7-7_S37_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16087 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey038-N7-8_S38_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16613 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey039-N7-9_S39_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8158 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey040-N7-10_S40_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16960 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey041-N7-11_S41_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6431 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey042-N9-2_S42_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16655 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey043-N9-3_S43_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10010 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey044-N9-4_S44_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14331 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey045-N9-5_S45_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12197 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey046-N9-6_S46_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17219 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey047-N9-7_S47_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6965 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey048-N9-8_S48_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17680 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey049-N9-9_S49_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6436 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey050-N9-10_S50_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16610 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey051-N9-11_S51_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15521 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey052-N11-2_S52_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15682 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey053-N11-3_S53_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17346 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey054-N11-4_S54_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12910 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey055-N11-5_S55_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13471 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey056-N11-6_S56_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21226 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey057-N11-7_S57_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14202 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey058-N11-8_S58_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21470 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey059-N11-9_S59_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18643 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey060-N11-10_S60_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 83166 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey061-N11-11_S61_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 68221 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey063-N13-2_S63_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13068 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey064-N13-3_S64_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16534 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey065-N13-4_S65_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10298 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey066-N13-5_S66_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12938 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey067-N13-6_S67_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14594 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey068-N13-7_S68_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13215 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey069-N13-8_S69_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 26642 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey070-N13-9_S70_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 40199 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey071-N13-10_S71_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 27210 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey072-N13-11_S72_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17023 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey074-N15-2_S74_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4796 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey075-N15-3_S75_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9618 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey076-N15-4_S76_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9085 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey077-N15-5_S77_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17793 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey078-N15-6_S78_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9776 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey079-N15-7_S79_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4109 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey080-N15-8_S80_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14473 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey081-N15-9_S81_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 24687 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey082-N15-10_S82_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 85727 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> 661Roey083-N15-11_S83_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 75419 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel085-N17-2_S85_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14554 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel086-N17-3_S86_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15569 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel087-N17-4_S87_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14161 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel088-N17-5_S88_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19344 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel089-N17-6_S89_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17271 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel090-N17-7_S90_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19219 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel091-N17-8_S91_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19414 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel092-N17-9_S92_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 80782 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel093-N17-10_S93_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 186685 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel094-N17-11_S94_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 144876 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel095-N19-2_S95_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6694 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel096-N19-3_S96_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17213 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel097-N19-4_S97_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8845 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel098-N19-5_S98_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4452 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel099-N19-6_S99_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10198 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel100-N19-7_S100_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13570 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel101-N19-8_S101_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12203 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel102-N19-9_S102_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18407 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel103-N19-10_S103_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19347 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel104-N19-11_S104_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13288 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel105-TG1-2_S105_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 101730 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel106-TG1-3_S106_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 130864 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel107-TG1-4_S107_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 167090 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel108-TG1-5_S108_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 215879 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel109-TG1-6_S109_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 269173 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel110-TG1-7_S110_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 246238 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel111-TG1-8_S111_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 215001 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel112-TG1-9_S112_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 227359 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel113-TG1-10_S113_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 163783 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel114-TG1-11_S114_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 179663 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel115-TG3-2_S115_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14517 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel116-TG3-3_S116_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 12077 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel117-TG3-4_S117_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16453 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel118-TG3-5_S118_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 34084 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel119-TG3-6_S119_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 125588 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel120-TG3-7_S120_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 102357 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel121-TG3-8_S121_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 167158 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel122-TG3-9_S122_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 127689 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel123-TG3-10_S123_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 76447 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel124-TG3-11_S124_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21357 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel125-TG5-2_S125_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18398 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel126-TG5-3_S126_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18979 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel127-TG5-4_S127_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14772 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel128-TG5-5_S128_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16446 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel129-TG5-6_S129_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 27429 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel130-TG5-7_S130_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 22246 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel131-TG5-8_S131_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17446 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel132-TG5-9_S132_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21592 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel133-TG5-10_S133_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 20832 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel134-TG5-11_S134_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13517 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel135-TG7-2_S135_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8037 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel136-TG7-3_S136_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 3633 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel137-TG7-4_S137_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 2214 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel138-TG7-5_S138_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 1692 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel139-TG7-6_S139_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 11821 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel140-TG7-7_S140_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 3910 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel141-TG7-8_S141_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6934 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel142-TG7-9_S142_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13078 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel143-TG7-10_S143_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 6471 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel144-TG7-11_S144_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 9046 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel145-TG9-2_S145_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13436 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel146-TG9-3_S146_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 21091 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel147-TG9-4_S147_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 7885 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel148-TG9-5_S148_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8122 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel149-TG9-6_S149_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 2293 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel150-TG9-7_S150_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 742 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel151-TG9-8_S151_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4430 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel152-TG9-9_S152_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 2851 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel153-TG9-10_S153_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 1410 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Angel154-TG9-11_S154_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 10878 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey155-TG11-2_S155_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4525 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey156-TG11-3_S156_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16456 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey157-TG11-4_S157_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17004 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey158-TG11-5_S158_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 32212 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey159-TG11-6_S159_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 28696 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey160-TG11-7_S160_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 22549 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey161-TG11-8_S161_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 24447 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey162-TG11-9_S162_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 33439 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey163-TG11-10_S163_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 34230 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey164-TG11-11_S164_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 13802 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey165-TG13-2_S165_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15564 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey166-TG13-3_S166_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15994 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey167-TG13-4_S167_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 18109 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey168-TG13-5_S168_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 11890 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey169-TG13-6_S169_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19300 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey170-TG13-7_S170_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16875 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey171-TG13-8_S171_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14456 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey172-TG13-9_S172_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16984 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey173-TG13-10_S173_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4571 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey174-TG13-11_S174_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16520 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey175-TG15-2_S175_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 8169 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey176-TG15-3_S176_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 16923 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey177-TG15-4_S177_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 27422 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey178-TG15-5_S178_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14417 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey179-TG15-6_S179_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 22639 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey180-TG15-7_S180_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 56232 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey181-TG15-8_S181_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 74640 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey182-TG15-9_S182_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 44146 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey183-TG15-10_S183_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 42247 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey184-TG15-11_S184_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 19432 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey185-TG17-2_S185_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15643 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey186-TG17-3_S186_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15140 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey187-TG17-4_S187_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14787 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey188-TG17-5_S188_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17634 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey189-TG17-6_S189_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 29522 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey190-TG17-7_S190_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 60698 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey191-TG17-8_S191_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 42807 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey192-TG17-9_S192_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 38477 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey193-TG17-10_S193_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 47505 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey194-TG17-11_S194_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14013 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey195-TG19-2_S195_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4229 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey196-TG19-3_S196_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 4144 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey197-TG19-4_S197_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 14124 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey198-TG19-5_S198_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 15355 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey199-TG19-6_S199_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 29635 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey200-TG19-7_S200_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 34411 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey201-TG19-8_S201_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 87011 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey202-TG19-9_S202_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 161094 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey203-TG19-10_S203_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 138238 </td>
  </tr>
  <tr>
   <td style="text-align:left;"> Roey204-TG19-11_S204_L001_R1_001_noPrimers.fastq.gz </td>
   <td style="text-align:right;"> 17607 </td>
  </tr>
</tbody>
</table>

```r
Ps_obj_ord <-
  ordinate(Ps_obj,
           "MDS",
           "bray")
plot_ordination(
  Ps_obj,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/plot abundance-4.svg)<!-- -->

```r
# geom_point(size = 5) +
# geom_text(size = 5) 
```
The figure and table indicate only a small deviation in the number of reads per samples.
The control samples look more or less OK and separate from the other samples.

I will test now the effect of library size, and more importantly of the Illumina platform type on the community composition and also plot 

```r
(mod1 <- adonis(phyloseq::distance(Ps_obj, method = "bray") ~ Library.size,
  data = get_variable(Ps_obj),
  permutations = 9999
))
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj, method = "bray") ~      Library.size, data = get_variable(Ps_obj), permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##               Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Library.size   1     7.768  7.7680  21.104 0.09632  1e-04 ***
## Residuals    198    72.881  0.3681         0.90368           
## Total        199    80.649                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
plot_read_dist(Ps_obj)
```

![](04_Normalisation_figures/mod abundance-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj)
```

![](04_Normalisation_figures/mod abundance-2.svg)<!-- -->
Modelling library size shows a significant effect of read depth on the community structure, but explaining only 10% of the variance.
The reads histogram shows as expected a highly sparse and skewed sequence matrix.
The mean vs SD also shows as expected large dependency of SD on the mean reads of a sequence across all samples.

### Try various normalisation methods


```r
# subsample libraries from 1000 to max(sample_sums(Ps_obj)) and test
for (i in seq(1000, max(sample_sums(Ps_obj)), 1000)) {
  min_seqs <<- i
  Ps_obj_pruned_harsh <- subset_samples(Ps_obj, sample_sums(Ps_obj) > i)
  mod <-
    adonis2(phyloseq::distance(Ps_obj_pruned_harsh, method = "bray") ~ sample_sums(Ps_obj_pruned_harsh), # I use adonis2 because it gives a data.frame
      data = as(sample_data(Ps_obj_pruned_harsh), "data.frame"),
      permutations = 9999
    )
  Pval <- tidy(mod)$p.value[1]
  if (Pval > 0.05)
    break()
}
```

Only by subsetting the samples to a minimum library size of $1.39\times 10^{5}$ sequences do we get independence from library size but this forces us to drop $183$ out of $200$ samples.

Let's see the effect of this

```r
adonis(
  phyloseq::distance(Ps_obj_pruned_harsh, method = "bray") ~ Library.size,
  data =
    get_variable(Ps_obj_pruned_harsh),
  permutations = 9999
)
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj_pruned_harsh, method = "bray") ~      Library.size, data = get_variable(Ps_obj_pruned_harsh), permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##              Df SumsOfSqs MeanSqs F.Model     R2 Pr(>F)  
## Library.size  1   0.37999 0.37999  2.0803 0.1218   0.06 .
## Residuals    15   2.73987 0.18266         0.8782         
## Total        16   3.11986                 1.0000         
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
# 
# test_mat <- phyloseq::distance(Ps_obj_pruned_harsh, method = "bray")
# adonis(
#   adonis_f,
#   data =
#     as(sample_data(Ps_obj_pruned_harsh), "data.frame"),
#   permutations = 999
# )

plot_read_dist(Ps_obj_pruned_harsh)
```

![](04_Normalisation_figures/effect s pruning-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_pruned_harsh)
```

![](04_Normalisation_figures/effect s pruning-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_harsh,
           "CAP",
           "bray",
           formula = Ps_obj_pruned_harsh ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_harsh,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/effect s pruning - ordinate-1.svg)<!-- -->

Instead let's drop all samples below 2500 reads and do try some correction methods for library depths

```r
Ps_obj_pruned_min <-
  subset_samples(Ps_obj, sample_sums(Ps_obj) > min_lib_size)
Ps_obj_pruned_min <-
  filter_taxa(Ps_obj_pruned_min, function(x)
    sum(x) > 0, TRUE)
```

#### Rarefaction

```r
Ps_obj_pruned_rared <-
  rarefy_even_depth(
    Ps_obj_pruned_min,
    sample.size = min(sample_sums(Ps_obj_pruned_min)),
    rngseed = FALSE,
    replace = FALSE
  )
sample_data(Ps_obj_pruned_rared)$Library.size <-
  sample_sums(Ps_obj_pruned_rared)
# Ps_obj_pruned_rared <- Ps_obj_pruned_min
# Ps_obj_pruned_min %>%
#   otu_table() %>%
#   rowSums() %>%
#   min() %>%
#   rrarefy(otu_table(Ps_obj_pruned_min), .) ->
#   otu_table(Ps_obj_pruned_rared)

test_mat <- phyloseq::distance(Ps_obj_pruned_rared, method = "bray")
adonis(
  adonis_f,
  data =
    as(sample_data(Ps_obj_pruned_rared), "data.frame"),
  permutations = 999
)
```

```
## 
## Call:
## adonis(formula = adonis_f, data = as(sample_data(Ps_obj_pruned_rared),      "data.frame"), permutations = 999) 
## 
## Permutation: free
## Number of permutations: 999
## 
## Terms added sequentially (first to last)
## 
##                        Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Label..18O.             1     2.137  2.1370  5.8139 0.02803  0.001 ***
## Treatment               1     3.316  3.3155  9.0203 0.04349  0.001 ***
## Label..18O.:Treatment   1     0.582  0.5816  1.5824 0.00763  0.037 *  
## Residuals             191    70.205  0.3676         0.92085           
## Total                 194    76.239                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
qplot(rowSums(otu_table(Ps_obj_pruned_rared)),binwidth = 100, geom = "histogram") +   xlab("Library size")
```

![](04_Normalisation_figures/test rarefaction-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_pruned_rared)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/test rarefaction-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_pruned_rared, Var1, Var2, Var3)
```

![](04_Normalisation_figures/test rarefaction-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_pruned_rared)
```

![](04_Normalisation_figures/test rarefaction diag plots-1.svg)<!-- -->

```r
notAllZero <- (rowSums(t(otu_table(Ps_obj_pruned_rared))) > 0)
plot_mean_SD(Ps_obj_pruned_rared)
```

![](04_Normalisation_figures/test rarefaction diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_rared,
           "CAP",
           "bray",
           formula = Ps_obj_pruned_rared ~  get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_rared,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/rarefaction - ordinate-1.svg)<!-- -->

#### Scale libraries 

```r
Ps_obj_pruned_scaled <- Ps_obj_pruned_min
Ps_obj_scaled <- scale_libraries(Ps_obj_pruned_min)
sample_data(Ps_obj_scaled)$Library.size <- sample_sums(Ps_obj_scaled)

adonis(
  phyloseq::distance(Ps_obj_scaled, method = "bray") ~ Library.size,
  data =
    as(sample_data(Ps_obj_scaled), "data.frame"),
  permutations = 9999
)
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj_scaled, method = "bray") ~      Library.size, data = as(sample_data(Ps_obj_scaled), "data.frame"),      permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##               Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Library.size   1     7.407  7.4069  20.781 0.09721  1e-04 ***
## Residuals    193    68.790  0.3564         0.90279           
## Total        194    76.197                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
test_mat <- phyloseq::distance(Ps_obj_scaled, method = "bray")
adonis(
  adonis_f,
  data =
    as(sample_data(Ps_obj_scaled), "data.frame"),
  permutations = 999
)
```

```
## 
## Call:
## adonis(formula = adonis_f, data = as(sample_data(Ps_obj_scaled),      "data.frame"), permutations = 999) 
## 
## Permutation: free
## Number of permutations: 999
## 
## Terms added sequentially (first to last)
## 
##                        Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Label..18O.             1     2.136  2.1362  5.8181 0.02804  0.001 ***
## Treatment               1     3.344  3.3445  9.1087 0.04389  0.001 ***
## Label..18O.:Treatment   1     0.586  0.5865  1.5973 0.00770  0.057 .  
## Residuals             191    70.129  0.3672         0.92037           
## Total                 194    76.197                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
qplot(rowSums(otu_table(Ps_obj_scaled)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/scale-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_scaled)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/scale-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_scaled, Var1, Var2, Var3)
```

![](04_Normalisation_figures/scale-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_scaled)
```

![](04_Normalisation_figures/scaled diag plots-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_scaled)
```

![](04_Normalisation_figures/scaled diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_scaled,
           "CAP",
           "bray",
           formula = Ps_obj_scaled ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_scaled,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/scaled - ordinate-1.svg)<!-- -->


#### Cumulative sum scaling normalization [@paulson_differential_2013]

```r
# Cumulative sum scaling normalization
Ps_obj_pruned_CS <- Ps_obj_pruned_min

Ps_obj_pruned_CS %>%
  otu_table(.) %>%
  t() %>%
  as(., "matrix") %>%
  newMRexperiment(.) ->
  mr_obj
p <- cumNormStatFast(mr_obj)
cumNormMat(mr_obj, p = p) %>% 
  otu_table(., taxa_are_rows = TRUE) ->
  otu_table(Ps_obj_pruned_CS)

# Did any OTU produce Na?
Ps_obj_pruned_CS %>% 
  otu_table() %>% 
  as(., "matrix") %>% 
  t(.) %>% 
  apply(., 2, function(x) !any(is.na(x))) %>% 
  unlist() ->
  OTUs2keep
Ps_obj_pruned_CS <- prune_taxa(OTUs2keep, Ps_obj_pruned_CS)
sample_data(Ps_obj_pruned_CS)$Library.size <- sample_sums(Ps_obj_pruned_CS)

# adonis(
#   phyloseq::distance(Ps_obj_pruned_CS, method = "bray") ~ Library.size,
#   data =
#     as(sample_data(Ps_obj_pruned_CS), "data.frame"),
#   permutations = 9999
# )
# adonis(
#   phyloseq::distance(Ps_obj_pruned_CS, method = "bray") ~  get(Var4) * get(Var3) * get(Var2),
#   data =
#     as(sample_data(Ps_obj_pruned_CS), "data.frame"),
#   permutations = 9999
# )

qplot(rowSums(otu_table(Ps_obj_pruned_CS)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/cumsum-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_pruned_CS)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/cumsum-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_pruned_CS, Var1, Var2, Var3)
```

![](04_Normalisation_figures/cumsum-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_pruned_CS)
```

![](04_Normalisation_figures/cumsum diag plots-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_pruned_CS)
```

![](04_Normalisation_figures/cumsum diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_CS,
           "CAP",
           "bray",
           formula = Ps_obj_pruned_CS ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_CS,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/CS - ordinate-1.svg)<!-- -->

#### Standardize abundances to the median sequencing depth (and convert to proportion)

```r
Ps_obj_pruned_min %>%
  otu_table(.) %>%
  as(., "matrix") %>%
  rowSums() %>% 
  median() ->
  total
standf = function(x, t=total) round(t * (x / sum(x)))
Ps_obj_pruned_median <- transform_sample_counts(Ps_obj_pruned_min, standf) # Standardize abundances to median sequencing depth
sample_data(Ps_obj_pruned_median)$Library.size <- sample_sums(Ps_obj_pruned_median)

adonis(
  phyloseq::distance(Ps_obj_pruned_median, method = "bray") ~ Library.size,
  data =
    as(sample_data(Ps_obj_pruned_median), "data.frame"),
  permutations = 9999
)
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj_pruned_median, method = "bray") ~      Library.size, data = as(sample_data(Ps_obj_pruned_median),      "data.frame"), permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##               Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Library.size   1     1.731 1.73061  4.4892 0.02273  1e-04 ***
## Residuals    193    74.403 0.38551         0.97727           
## Total        194    76.134                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
test_mat <- phyloseq::distance(Ps_obj_pruned_median, method = "bray")
adonis(
  adonis_f,
  data =
    as(sample_data(Ps_obj_pruned_median), "data.frame"),
  permutations = 999
)
```

```
## 
## Call:
## adonis(formula = adonis_f, data = as(sample_data(Ps_obj_pruned_median),      "data.frame"), permutations = 999) 
## 
## Permutation: free
## Number of permutations: 999
## 
## Terms added sequentially (first to last)
## 
##                        Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Label..18O.             1     2.132  2.1322  5.8124 0.02801  0.001 ***
## Treatment               1     3.348  3.3483  9.1275 0.04398  0.001 ***
## Label..18O.:Treatment   1     0.587  0.5870  1.6002 0.00771  0.035 *  
## Residuals             191    70.066  0.3668         0.92030           
## Total                 194    76.134                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
qplot(rowSums(otu_table(Ps_obj_pruned_median)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/median-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_pruned_median)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/median-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_pruned_median, Var1, Var2, Var3)
```

![](04_Normalisation_figures/median-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_pruned_median)
```

![](04_Normalisation_figures/median diag plots-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_pruned_median)
```

![](04_Normalisation_figures/median diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_median,
           "CAP",
           "bray",
           formula = Ps_obj_pruned_median ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_median,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/median - ordinate-1.svg)<!-- -->

#### Standardize abundances using log transformation for variance stabilisation

```r
Ps_obj_pruned_log <- transform_sample_counts(Ps_obj_pruned_min, function(x) log(1 + x))

# Ps_obj_pruned_rlog <- Ps_obj_pruned_min
# Ps_obj_pruned_min %>%
#   transform_sample_counts(., function(x) (1 + x)) %>% # add pseudocount
#   phyloseq_to_deseq2(., ~ Spill) %>%
#   rlog(., blind = TRUE , fitType = "parametric") %>%
#   assay() %>%
#   otu_table(, taxa_are_rows = TRUE) ->
#   otu_table(Ps_obj_pruned_rlog)
sample_data(Ps_obj_pruned_log)$Library.size <- sample_sums(Ps_obj_pruned_log)

adonis(
  phyloseq::distance(Ps_obj_pruned_log, method = "bray") ~ Library.size,
  data =
    as(sample_data(Ps_obj_pruned_log), "data.frame"),
  permutations = 9999
)
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj_pruned_log, method = "bray") ~      Library.size, data = as(sample_data(Ps_obj_pruned_log), "data.frame"),      permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##               Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Library.size   1     8.042  8.0418  24.206 0.11144  1e-04 ***
## Residuals    193    64.119  0.3322         0.88856           
## Total        194    72.161                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
test_mat <- phyloseq::distance(Ps_obj_pruned_log, method = "bray")
adonis(
  adonis_f,
  data =
    as(sample_data(Ps_obj_pruned_log), "data.frame"),
  permutations = 999
)
```

```
## 
## Call:
## adonis(formula = adonis_f, data = as(sample_data(Ps_obj_pruned_log),      "data.frame"), permutations = 999) 
## 
## Permutation: free
## Number of permutations: 999
## 
## Terms added sequentially (first to last)
## 
##                        Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Label..18O.             1     1.983  1.9833  5.7337 0.02748  0.001 ***
## Treatment               1     3.475  3.4752 10.0470 0.04816  0.001 ***
## Label..18O.:Treatment   1     0.636  0.6356  1.8375 0.00881  0.018 *  
## Residuals             191    66.067  0.3459         0.91555           
## Total                 194    72.161                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
qplot(rowSums(otu_table(Ps_obj_pruned_log)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/log-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_pruned_log)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/log-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_pruned_log, Var1, Var2, Var3)
```

![](04_Normalisation_figures/log-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_pruned_log)
```

![](04_Normalisation_figures/log diag plots-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_pruned_log)
```

![](04_Normalisation_figures/log diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_log,
           "CAP",
           "bray",
           formula = Ps_obj_pruned_log ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_log,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/log - ordinate-1.svg)<!-- -->

#### Standardize abundances using Centered-Log-Ratio transformation for variance stabilisation [@fernandes_anova-like_2013]

```r
Ps_obj_pruned_CLR <- Ps_obj_pruned_min
Ps_obj_pruned_CLR <- phyloseq_CLR(Ps_obj_pruned_min)
```

```
## No. corrected values:  66856
```

```r
sample_data(Ps_obj_pruned_CLR)$Library.size <- sample_sums(Ps_obj_pruned_CLR)

adonis(
  phyloseq::distance(Ps_obj_pruned_CLR, method = "euclidean") ~ Library.size,
  data =
    as(sample_data(Ps_obj_pruned_CLR), "data.frame"),
  permutations = 9999
)
```

```
## 
## Call:
## adonis(formula = phyloseq::distance(Ps_obj_pruned_CLR, method = "euclidean") ~      Library.size, data = as(sample_data(Ps_obj_pruned_CLR), "data.frame"),      permutations = 9999) 
## 
## Permutation: free
## Number of permutations: 9999
## 
## Terms added sequentially (first to last)
## 
##               Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)
## Library.size   1      4249  4248.8 0.75019 0.00387 0.8574
## Residuals    193   1093075  5663.6         0.99613       
## Total        194   1097324                 1.00000
```

```r
test_mat <- phyloseq::distance(Ps_obj_pruned_CLR, method = "euclidean")
adonis(
  adonis_f,
  data =
    as(sample_data(Ps_obj_pruned_CLR), "data.frame"),
  permutations = 999
)
```

```
## 
## Call:
## adonis(formula = adonis_f, data = as(sample_data(Ps_obj_pruned_CLR),      "data.frame"), permutations = 999) 
## 
## Permutation: free
## Number of permutations: 999
## 
## Terms added sequentially (first to last)
## 
##                        Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
## Label..18O.             1     15387   15387  2.8646 0.01402  0.001 ***
## Treatment               1     46640   46640  8.6831 0.04250  0.001 ***
## Label..18O.:Treatment   1      9358    9358  1.7421 0.00853  0.030 *  
## Residuals             191   1025939    5371         0.93495           
## Total                 194   1097324                 1.00000           
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

```r
qplot(rowSums(otu_table(Ps_obj_pruned_CLR)), geom = "histogram") + 
  xlab("Library size")
```

![](04_Normalisation_figures/clr-1.svg)<!-- -->

```r
qplot(log10(rowSums(otu_table(Ps_obj_pruned_CLR)))) +
  xlab("Logged library size")
```

![](04_Normalisation_figures/clr-2.svg)<!-- -->

```r
plot_lib_size(Ps_obj_pruned_CLR, Var1, Var2, Var3)
```

![](04_Normalisation_figures/clr-3.svg)<!-- -->

```r
plot_read_dist(Ps_obj_pruned_CLR)
```

![](04_Normalisation_figures/clr diag plots-1.svg)<!-- -->

```r
plot_mean_SD(Ps_obj_pruned_CLR)
```

![](04_Normalisation_figures/clr diag plots-2.svg)<!-- -->


```r
Ps_obj_ord <-
  ordinate(Ps_obj_pruned_CLR,
           "CAP",
           "euclidean",
           formula = Ps_obj_pruned_CLR ~ get(Var3) * get(Var2))
plot_ordination(
  Ps_obj_pruned_CLR,
  Ps_obj_ord,
  type = "samples",
  color = Var2,
  # label = Var4,
  shape = Var3
)  + 
  geom_text(mapping = aes(label = get(Var2)), size = 4, alpha = 3/4) +
  geom_point(size = 4)
```

![](04_Normalisation_figures/clr - ordinate-1.svg)<!-- -->


```r
sessioninfo::session_info() %>%
  details::details(
    summary = 'Current session info',
    open    = TRUE
  )
```

<details open>
<summary> <span title='Click to Expand'> Current session info </span> </summary>

```r

─ Session info ─────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.1.1 (2021-08-10)
 os       Ubuntu 18.04.6 LTS          
 system   x86_64, linux-gnu           
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Prague               
 date     2021-10-21                  

─ Packages ─────────────────────────────────────────────────────────────────────────────
 package          * version  date       lib source        
 ade4               1.7-18   2021-09-16 [1] CRAN (R 4.1.1)
 affy               1.70.0   2021-05-19 [1] Bioconductor  
 affyio             1.62.0   2021-05-19 [1] Bioconductor  
 ape                5.5      2021-04-25 [1] CRAN (R 4.0.3)
 assertthat         0.2.1    2019-03-21 [1] CRAN (R 4.0.2)
 backports          1.2.1    2020-12-09 [1] CRAN (R 4.0.2)
 Biobase          * 2.52.0   2021-05-19 [1] Bioconductor  
 BiocGenerics     * 0.38.0   2021-05-19 [1] Bioconductor  
 BiocManager        1.30.16  2021-06-15 [1] CRAN (R 4.1.0)
 biomformat         1.20.0   2021-05-19 [1] Bioconductor  
 Biostrings         2.60.2   2021-08-05 [1] Bioconductor  
 bitops             1.0-7    2021-04-24 [1] CRAN (R 4.0.3)
 broom            * 0.7.9    2021-07-27 [1] CRAN (R 4.1.0)
 bslib              0.3.1    2021-10-06 [1] CRAN (R 4.1.1)
 caTools            1.18.2   2021-03-28 [1] CRAN (R 4.0.3)
 cellranger         1.1.0    2016-07-27 [1] CRAN (R 4.0.2)
 cli                3.0.1    2021-07-17 [1] CRAN (R 4.1.0)
 clipr              0.7.1    2020-10-08 [1] CRAN (R 4.0.2)
 cluster            2.1.2    2021-04-17 [1] CRAN (R 4.0.3)
 codetools          0.2-18   2020-11-04 [1] CRAN (R 4.0.2)
 colorspace         2.0-2    2021-06-24 [1] CRAN (R 4.1.0)
 crayon             1.4.1    2021-02-08 [1] CRAN (R 4.0.3)
 data.table         1.14.2   2021-09-27 [1] CRAN (R 4.1.1)
 DBI                1.1.1    2021-01-15 [1] CRAN (R 4.0.3)
 dbplyr             2.1.1    2021-04-06 [1] CRAN (R 4.0.3)
 desc               1.4.0    2021-09-28 [1] CRAN (R 4.1.1)
 details            0.2.1    2020-01-12 [1] CRAN (R 4.0.2)
 digest             0.6.28   2021-09-23 [1] CRAN (R 4.1.1)
 dplyr            * 1.0.7    2021-06-18 [1] CRAN (R 4.1.0)
 ellipsis           0.3.2    2021-04-29 [1] CRAN (R 4.0.3)
 evaluate           0.14     2019-05-28 [1] CRAN (R 4.0.2)
 extrafont        * 0.17     2014-12-08 [1] CRAN (R 4.1.0)
 extrafontdb        1.0      2012-06-11 [1] CRAN (R 4.0.2)
 fansi              0.5.0    2021-05-25 [1] CRAN (R 4.0.3)
 farver             2.1.0    2021-02-28 [1] CRAN (R 4.0.3)
 fastmap            1.1.0    2021-01-25 [1] CRAN (R 4.0.3)
 forcats          * 0.5.1    2021-01-27 [1] CRAN (R 4.0.3)
 foreach            1.5.1    2020-10-15 [1] CRAN (R 4.0.2)
 fs                 1.5.0    2020-07-31 [1] CRAN (R 4.0.2)
 generics           0.1.0    2020-10-31 [1] CRAN (R 4.0.2)
 GenomeInfoDb       1.28.4   2021-09-05 [1] Bioconductor  
 GenomeInfoDbData   1.2.6    2021-05-25 [1] Bioconductor  
 ggplot2          * 3.3.5    2021-06-25 [1] CRAN (R 4.1.0)
 glmnet           * 4.1-2    2021-06-24 [1] CRAN (R 4.1.0)
 glue               1.4.2    2020-08-27 [1] CRAN (R 4.0.2)
 gplots             3.1.1    2020-11-28 [1] CRAN (R 4.0.2)
 gtable             0.3.0    2019-03-25 [1] CRAN (R 4.0.2)
 gtools             3.9.2    2021-06-06 [1] CRAN (R 4.1.0)
 haven              2.4.3    2021-08-04 [1] CRAN (R 4.1.0)
 hexbin             1.28.2   2021-01-08 [1] CRAN (R 4.0.2)
 highr              0.9      2021-04-16 [1] CRAN (R 4.0.3)
 hms                1.1.1    2021-09-26 [1] CRAN (R 4.1.1)
 htmltools          0.5.2    2021-08-25 [1] CRAN (R 4.1.1)
 httr               1.4.2    2020-07-20 [1] CRAN (R 4.0.2)
 igraph             1.2.7    2021-10-15 [1] CRAN (R 4.1.1)
 IRanges            2.26.0   2021-05-19 [1] Bioconductor  
 iterators          1.0.13   2020-10-15 [1] CRAN (R 4.0.2)
 jquerylib          0.1.4    2021-04-26 [1] CRAN (R 4.0.3)
 jsonlite           1.7.2    2020-12-09 [1] CRAN (R 4.0.2)
 kableExtra       * 1.3.4    2021-02-20 [1] CRAN (R 4.0.3)
 KernSmooth         2.23-20  2021-05-03 [1] CRAN (R 4.0.3)
 knitr              1.36     2021-09-29 [1] CRAN (R 4.1.1)
 labeling           0.4.2    2020-10-20 [1] CRAN (R 4.0.2)
 lattice          * 0.20-45  2021-09-22 [1] CRAN (R 4.1.1)
 lifecycle          1.0.1    2021-09-24 [1] CRAN (R 4.1.1)
 limma            * 3.48.3   2021-08-10 [1] Bioconductor  
 locfit             1.5-9.4  2020-03-25 [1] CRAN (R 4.0.2)
 lubridate          1.8.0    2021-10-07 [1] CRAN (R 4.1.1)
 magrittr         * 2.0.1    2020-11-17 [1] CRAN (R 4.0.2)
 MASS               7.3-54   2021-05-03 [1] CRAN (R 4.0.3)
 Matrix           * 1.3-4    2021-06-01 [1] CRAN (R 4.1.0)
 matrixStats        0.61.0   2021-09-17 [1] CRAN (R 4.1.1)
 metagenomeSeq    * 1.34.0   2021-05-19 [1] Bioconductor  
 mgcv               1.8-38   2021-10-06 [1] CRAN (R 4.1.1)
 modelr             0.1.8    2020-05-19 [1] CRAN (R 4.0.2)
 multtest           2.48.0   2021-05-19 [1] Bioconductor  
 munsell            0.5.0    2018-06-12 [1] CRAN (R 4.0.2)
 NADA               1.6-1.1  2020-03-22 [1] CRAN (R 4.0.2)
 nlme               3.1-153  2021-09-07 [1] CRAN (R 4.1.1)
 permute          * 0.9-5    2019-03-12 [1] CRAN (R 4.0.2)
 phyloseq         * 1.36.0   2021-05-19 [1] Bioconductor  
 pillar             1.6.4    2021-10-18 [1] CRAN (R 4.1.1)
 pkgconfig          2.0.3    2019-09-22 [1] CRAN (R 4.0.2)
 plyr               1.8.6    2020-03-03 [1] CRAN (R 4.0.2)
 png                0.1-7    2013-12-03 [1] CRAN (R 4.0.2)
 preprocessCore     1.54.0   2021-05-19 [1] Bioconductor  
 purrr            * 0.3.4    2020-04-17 [1] CRAN (R 4.0.2)
 R6                 2.5.1    2021-08-19 [1] CRAN (R 4.1.1)
 ragg             * 1.1.3    2021-06-09 [1] CRAN (R 4.1.0)
 RColorBrewer     * 1.1-2    2014-12-07 [1] CRAN (R 4.0.2)
 Rcpp               1.0.7    2021-07-07 [1] CRAN (R 4.1.0)
 RCurl              1.98-1.5 2021-09-17 [1] CRAN (R 4.1.1)
 readr            * 2.0.2    2021-09-27 [1] CRAN (R 4.1.1)
 readxl             1.3.1    2019-03-13 [1] CRAN (R 4.0.2)
 reprex             2.0.1    2021-08-05 [1] CRAN (R 4.1.0)
 reshape2           1.4.4    2020-04-09 [1] CRAN (R 4.0.2)
 rhdf5              2.36.0   2021-05-19 [1] Bioconductor  
 rhdf5filters       1.4.0    2021-05-19 [1] Bioconductor  
 Rhdf5lib           1.14.2   2021-07-06 [1] Bioconductor  
 rlang              0.4.12   2021-10-18 [1] CRAN (R 4.1.1)
 rmarkdown        * 2.11     2021-09-14 [1] CRAN (R 4.1.1)
 rprojroot          2.0.2    2020-11-15 [1] CRAN (R 4.0.2)
 rstudioapi         0.13     2020-11-12 [1] CRAN (R 4.0.2)
 Rttf2pt1           1.3.9    2021-07-22 [1] CRAN (R 4.1.0)
 rvest              1.0.2    2021-10-16 [1] CRAN (R 4.1.1)
 S4Vectors          0.30.2   2021-10-03 [1] Bioconductor  
 sass               0.4.0    2021-05-12 [1] CRAN (R 4.0.3)
 scales           * 1.1.1    2020-05-11 [1] CRAN (R 4.0.2)
 sessioninfo        1.1.1    2018-11-05 [1] CRAN (R 4.0.2)
 shape              1.4.6    2021-05-19 [1] CRAN (R 4.0.3)
 stringi            1.7.5    2021-10-04 [1] CRAN (R 4.1.1)
 stringr          * 1.4.0    2019-02-10 [1] CRAN (R 4.0.2)
 survival           3.2-13   2021-08-24 [1] CRAN (R 4.1.1)
 svglite          * 2.0.0    2021-02-20 [1] CRAN (R 4.1.0)
 systemfonts        1.0.3    2021-10-13 [1] CRAN (R 4.1.1)
 textshaping        0.3.6    2021-10-13 [1] CRAN (R 4.1.1)
 tibble           * 3.1.5    2021-09-30 [1] CRAN (R 4.1.1)
 tidyr            * 1.1.4    2021-09-27 [1] CRAN (R 4.1.1)
 tidyselect         1.1.1    2021-04-30 [1] CRAN (R 4.0.3)
 tidyverse        * 1.3.1    2021-04-15 [1] CRAN (R 4.0.3)
 truncnorm          1.0-8    2018-02-27 [1] CRAN (R 4.0.2)
 tzdb               0.1.2    2021-07-20 [1] CRAN (R 4.1.0)
 utf8               1.2.2    2021-07-24 [1] CRAN (R 4.1.0)
 vctrs              0.3.8    2021-04-29 [1] CRAN (R 4.0.3)
 vegan            * 2.5-7    2020-11-28 [1] CRAN (R 4.0.3)
 viridisLite        0.4.0    2021-04-13 [1] CRAN (R 4.0.3)
 vsn              * 3.60.0   2021-05-19 [1] Bioconductor  
 webshot            0.5.2    2019-11-22 [1] CRAN (R 4.0.2)
 withr              2.4.2    2021-04-18 [1] CRAN (R 4.0.3)
 Wrench             1.10.0   2021-05-19 [1] Bioconductor  
 xfun               0.27     2021-10-18 [1] CRAN (R 4.1.1)
 xml2               1.3.2    2020-04-23 [1] CRAN (R 4.0.2)
 XVector            0.32.0   2021-05-19 [1] Bioconductor  
 yaml               2.2.1    2020-02-01 [1] CRAN (R 4.0.2)
 zCompositions      1.3.4    2020-03-04 [1] CRAN (R 4.0.2)
 zlibbioc           1.38.0   2021-05-19 [1] Bioconductor  

[1] /home/angel/R/library
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library

```

</details>
<br>

```r
## References
```
